### PR TITLE
Add basic kotlin support

### PR DESCRIPTION
**Pull request by jaresty**
- Borrowed from [here](https://github.com/okonomichiyaki/knausj_talon/commit/6542dc38e43b591c2ca2ea1d1360ac198e6eb168)

Addresses https://github.com/talonhub/community/issues/1406

### CODE REVIEW

File: lang/kotlin/kotlin.talon [hunk 1]
```diff
@@ -0,0 +1,27 @@
+tag: user.kotlin
```
Comment (auscompgeek) [inline lang/kotlin/kotlin.talon:1]: This seems to predate #1256.

```suggestion
code.language: kotlin
```

File: lang/kotlin/kotlin.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+class UserActions:
+    def code_comment_line_prefix():
+        actions.insert("// ")
+
+    def code_operator_indirection():
+        actions.skip()
+
+    def code_operator_address_of():
+        actions.skip()
+
```
Comment (auscompgeek) [inline lang/kotlin/kotlin.py:19]: Actions shouldn't be implemented if they're not implementable. This suppresses the warning about the action being unimplemented when used. We don't `actions.skip()` anywhere else in `lang/`.

```suggestion
```

File: lang/kotlin/kotlin.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        actions.key("left")
+
+    def code_state_else():
+        actions.insert("else ")
+        actions.key("enter")
+
+    def code_state_switch():
+        actions.insert("switch () ")
+        actions.key("left")
+        actions.edit.left()
```
Comment (auscompgeek) [inline lang/kotlin/kotlin.py:137]: All of these which put the cursor between the `()` should use `actions.user.insert_between()`.

File: lang/kotlin/kotlin.talon (hunk 1 above)
Comment (jaresty) [reply lang/kotlin/kotlin.talon]: Fixed! Thanks.

File: lang/kotlin/kotlin.py (hunk 2 above)
Comment (jaresty) [reply lang/kotlin/kotlin.py]: Removed.

File: lang/kotlin/kotlin.py (hunk 3 above)
Comment (jaresty) [reply lang/kotlin/kotlin.py]: Updated these.

**Comment by auscompgeek** [comment]
Possibly silly question: how similar is Kotlin to Java? Would it possibly make sense for our Kotlin support to extend our Java support?

**Comment by jaresty** [comment]
Kotlin is pretty different from Java when you look at it closely. I don't think we'd want to borrow syntax from Java most of the time. For example, Kotlin does not require semicolons and Kotlin is more functional than Java. I think it's better to start with something independent at first.

File: lang/kotlin/kotlin.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+class UserActions:
+    def code_comment_line_prefix():
+        actions.insert("// ")
+
+    def code_operator_lambda():
+        actions.auto_insert(" -> ")
+
+    def code_operator_subscript():
+        actions.insert("[]")
+        actions.key("left")
```
Comment (okonomichiyaki) [inline lang/kotlin/kotlin.py:19]: should this also be using `actions.user.insert_between` ?

**Comment by okonomichiyaki** [comment]
thanks @jaresty for shepherding this into the community repo, and to auscompgeek for comments. I am neither a Kotlin expert nor up to date on the Talon API and patterns. have only used these files for a little Kotlin programming over the past year and was enough for basic stuff but I'm sure can be improved to support more of the language

File: lang/kotlin/kotlin.talon [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+settings():
+    user.code_private_function_formatter = "PRIVATE_CAMEL_CASE"
+    user.code_protected_function_formatter = "PRIVATE_CAMEL_CASE"
+    user.code_public_function_formatter = "PRIVATE_CAMEL_CASE"
+    user.code_private_variable_formatter = "PRIVATE_CAMEL_CASE"
+    user.code_protected_variable_formatter = "PRIVATE_CAMEL_CASE"
+    user.code_public_variable_formatter = "PRIVATE_CAMEL_CASE"
+
+state var: "var "
+state val: "val "
```
[the remark this answers is not in the document]
Comment (jaresty) [reply lang/kotlin/kotlin.talon]: This sounds like a good idea. I borrowed the modifiers from java but I'm not sure they're correct.

File: lang/kotlin/kotlin.py (hunk 4 above)
Comment (jaresty) [reply lang/kotlin/kotlin.py]: Oops I missed that one. Fixed now.

File: lang/kotlin/kotlin.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    "synchronized": "synchronized",
+    "volatile": "volatile",
+    "transient": "transient",
+    "abstract": "abstract",
+    "interface": "interface",
+    "final": "final",
+}
+
+mod.list("kotlin_modifier", desc="Kotlin Modifiers")
+ctx.lists["self.kotlin_modifier"] = kotlin_modifiers
```
Comment (knausj85) [inline lang/kotlin/kotlin.py:27]: I think we should probably modernize this to use the code_keywords stuff? See JavaScript for an example 

https://github.com/talonhub/community/blob/main/lang/tags/keywords.talon

**Review by knausj85 — CHANGES_REQUESTED** [review]

**Review by knausj85 — APPROVED** [review]

**Pull request merged**